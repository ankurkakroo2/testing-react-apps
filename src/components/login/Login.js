import React, { useState } from "react";

import { useFormInput, setUserSession } from "../../util/Common";
import {
  UserName,
  Password,
  Wrong_Details,
  Blank_UserName,
  Blank_Password
} from "../../constants";
import "./default.css";

function Login({ history }) {
  const username = useFormInput("");
  const password = useFormInput("");
  const [error, setError] = useState(null);
  const [userNameError, setUserNameError] = useState(null);
  const [passwordError, setPasswordError] = useState(null);
  const [loading, setLoading] = useState(false);

  const handleLogin = event => {
    event.preventDefault();
    resetErrors();
    if (validateUserDetails()) {
      setLoading(true);
      if (username.value === UserName && password.value === Password) {
        setUserSession(`${UserName}_${Math.random(0, 100)}`, username.value);
        history.push("/dashboard");
        setError(null);
        return;
      }

      setError(Wrong_Details);
      setLoading(false);
    }
  };
  const resetErrors = () => {
    setError(null);
    setUserNameError(null);
    setPasswordError(null);
  };
  const validateUserDetails = () => {
    if (!username.value.trim()) {
      setUserNameError(Blank_UserName);
      return false;
    } else if (!password.value.trim()) {
      setPasswordError(Blank_Password);
      return false;
    }
    setError(null);
    return true;
  };
  return (
    <section className="login-form">
      <h1 className="sign-in">Login</h1>
      <form onSubmit={handleLogin}>
        <div>
          <label htmlFor="username">Username</label>
          <input
            type="text"
            {...username}
            id="username"
            data-testid="username"
          />
        </div>
        {userNameError && (
          <p className="login-form__username-error">{userNameError}</p>
        )}
        <div>
          <label htmlFor="password">Password</label>
          <input
            type="password"
            {...password}
            id="password"
            data-testid="password"
          />
        </div>
        {passwordError && (
          <p className="login-form__password-error">{passwordError}</p>
        )}
        {error && <p className="login-form__error">{error}</p>}
        <input
          type="submit"
          value={loading ? "Loading..." : "Login"}
          disabled={loading}
          className="submit-btn"
          data-testid="submit-btn"
        />
      </form>
    </section>
  );
}

export default Login;
