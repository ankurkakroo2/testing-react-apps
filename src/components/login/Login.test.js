import React from "react";
import Login from "./index";
import { render, cleanup } from "@testing-library/react";

let result;
beforeEach(() => {
  result = render(<Login />);
});

afterEach(cleanup);

describe("Login", () => {
  test("render without crash", () => {
    expect(result.getByTestId("login-heading")).toBeInTheDocument();
  });

  test("snapshot", () => {
    expect(result.asFragment()).toMatchSnapshot();
  });
});

describe("UI tests", () => {
  test("render username field", () => {
    expect(result.getByLabelText("Username")).toBeInTheDocument();
  });
  test("render password field", () => {
    expect(result.getByLabelText("Password")).toBeInTheDocument();
  });
  test("render submit button", () => {
    expect(result.getByRole("button")).toBeInTheDocument();
  });
});
