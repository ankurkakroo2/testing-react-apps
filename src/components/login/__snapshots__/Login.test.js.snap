// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login snapshot 1`] = `
<DocumentFragment>
  <section
    class="login-form"
  >
    <h1
      class="sign-in"
    >
      Login
    </h1>
    <form>
      <div>
        <label
          for="username"
        >
          Username
        </label>
        <input
          data-test-id="username"
          id="username"
          type="text"
          value=""
        />
      </div>
      <div>
        <label
          for="password"
        >
          Password
        </label>
        <input
          data-test-id="password"
          id="password"
          type="password"
          value=""
        />
      </div>
      <input
        class="submit-btn"
        data-test-id="submit-btn"
        type="submit"
        value="Login"
      />
    </form>
  </section>
</DocumentFragment>
`;
